### ISSUE TITLE
Append key/value to posts to be used inside the resolve function.

### ISSUE THREAD

**Issue by a-barbieri** [wp-graphql/wp-graphql#1205]
This is a tricky one.

# TL;DR

I need to manually pass a **value** to a **field** as this is accessible only within the `RelevanssiConnectionResolver::get_query()` method. In other words the value cannot be retrieved with `get_post_meta()` method as you do on the `resolve` function of the `register_graphql_field` method. See [guidelines](https://docs.wpgraphql.com/extending/fields/#register-fields-to-the-schema).

# Details

I've been developing a Relevanssi plugin for WP GraphQL. You can find my [first steps here](https://github.com/wp-graphql/wp-graphql/issues/1185) and a [gist here](https://gist.github.com/a-barbieri/7802dee5938fee53e6f1a5cbd6f81cca)

The `relevanssi_do_query()` method – required to perform a Relevanssi search – returns a `relevance_score` key/value which needs to be accessible from GraphQL query. This field isn't persisted in the database, but it's generated whenever  `relevanssi_do_query()` is executed.

The value is accessible in the following way:

```php
$search_results = relevanssi_do_query( $query );
var_dump( $search_results[0]->relevance_score )

//   float(1124.2)
```

I was thinking about adding that value manually to each post before returning the query at the end of the `get_query()` method (PLEASE SEE [THE CODE](https://gist.github.com/a-barbieri/7802dee5938fee53e6f1a5cbd6f81cca#file-relevanssi-connection-resolver-php-L30) I'M REFERRING TO), but I don't think that's doable.

```php
get_query() {
    // ...some code

    // The updated query
    $updated_query = new \WP_Query( $this->query_args );

    // Loop through all posts in $updated_query 
    // and add to each one the relevance_score value.
    ????

    // return updated query enhanced
    return $updated_query;
}
```

Then I was thinking about retrieving `relevance_score` this way:

```php
register_graphql_field(
  'MyCustomPostType',
  'relevance_score',
  [
    'type'        => 'String',
    'description' => __( 'Relevanssi score based on current search query', 'wp-graphql' ),
    'resolve'     => function( $post ) {

      // I'd love to do this.
      $relevance_score = $post->relevance_score;

      return is_float($relevance_score) ? $relevance_score : null;
    },
  ]
);
```

How can I pass/append `relevance_score` to each post returned from the connection? Is it possible?

**Comment by jasonbahl** [comment]
@a-barbieri so, you wouldn't want the relevance score on the Post object itself, because the relevance is in relation to the specific query the Post is part of. 

Exposing the field on the Post would lead to potential cache issues. 

If you had 2 queries on the same page, and Post A was a result of both queries, a cache like Apollo will cache Post A one time, so the `relevanceScore` would be overridden by the Post in the later query. 

You want something like `relevanceScore` to live on the edge in the query. 

Edges are for contextual information in relation to a specific query. 

This way folks could query like so: 

```
{
   posts( where: { search: "some keyword" } ) {
      edges {
           relevanceScore
           node {
               id
               title
           }
       }
   }
}
```

This way, if 2 queries were executed at the same time with different searches, the relevance score could be displayed for each node without being overridden in a normalized cache. 

To add edge fields, you can do similar to what you have above, using the `register_graphql_field` method, but change the `MyCustomPostType` to the name of the Edge to add to. 

For example:

```
register_graphql_field( 'RootQueryToPostConnectionEdge', 'relevanceScore', [
  'type' => 'Float', // I'm assuming it's a float?
  'description' => __( 'Relevanssi score based on current search query', 'wp-graphql-relevanssi' ),
  'resolve' => function( $edge ) {
     // The $edge contains the source, cursor and node
     // Also, as of this PR (https://github.com/wp-graphql/wp-graphql/pull/1212), the edge also contains the connection
    // so we can pluck data out of what's passed down to resolve

    // This will be an instance of the PostObjectConnectionResolver class
    $connection = isset( $edge['connection'] ) ? $edge['connection'] : null;

    // This will be an instance of WPGraphQL\Model\Post();
    $node = isset( $edge['node'] ) ? $edge['node'] : null;

    // This will be the base64encoded cursor
    $cursor = isset( $edge['cursor'] ) ? $edge['cursor'] : null;

     // This will be the source. On a RootQuery connection, the source would be null
     $source = isset( $edge['source'] ) ? $edge['source'] : null; 

    // At this point, you should have enough context to get the relevance score from the $edge that's passed down and return it here.

  } 
]);
```

**Comment by jasonbahl** [comment]
Note, the PR I mentioned in my code comment (https://github.com/wp-graphql/wp-graphql/pull/1212)

**Comment by jasonbahl** [comment]
I'm thinking we also might want to introduce an API for registering edge fields to make it a bit easier. 

Right now, you have to have the full Edge name, like `RootQueryToPostConnectionEdge` and that's cumbersome. 

Might be nice to have something like: `register_graphql_edge_field( $fromType, $toType, $fieldName, $config );`

I think it's easier to know the from/to type than know the fully qualified edge name 🤔

**Comment by a-barbieri** [comment]
Wow, I wasn't thinking about cache at all; this is definitely good to know.

One little doubt `RootQueryToPostConnectionEdge` should be `RootQueryToDocumentConnectionEdge` in case I have a `Document` custom post type?

## My first try would be...

Going to the core of the issue, you are suggesting to append my data directly to the resolver instance (of the class `RelevanssiConnectionResolver`) which can be retrieved through `$edge->connection`, right? (based on your [code](https://github.com/jasonbahl/wp-graphql/blob/5b842008375d7934272bbb5d5c179be95652033e/src/Data/Connection/AbstractConnectionResolver.php#L589))

```php
get_query() {
    // ...some code

    // Get results
    $search_results = relevanssi_do_query( $query );

    // Get an array of IDs and scores
    $ids = wp_list_pluck( $search_results, 'ID' );
    $scores = wp_list_pluck( $search_results, 'relevance_score' );

    // Create an array to store ids and scores
    $relevance_scores = array();
    for ( $i = 0; $i < count($ids); $i++ ) {
      $current_id = $ids[$i];
      $relevance_scores[$current_id] = $scores[$i];
    }

    // Append this to be able to retrieve it later using `$edge->connection`
    $this->relevance_scores = $relevance_scores;

    // For example:
    // $relevance_scores = array(
    //     1283 => 1223.4,
    //     1094 => 234.1
    // );
    
    // ... rest of code
}
```

Then I should be able to do:

```php
register_graphql_field( 'RootQueryToPostConnectionEdge', 'relevanceScore', [
  'type' => 'Float', // yes, it's a float
  'description' => __( 'Relevanssi score based on current search query', 'wp-graphql-relevanssi' ),
  'resolve' => function( $edge ) {
    $connection = isset( $edge['connection'] ) ? $edge['connection'] : null;
    $node = isset( $edge['node'] ) ? $edge['node'] : null;
    if ( is_null($connection) || is_null($node) ) {
      return null;
    } else {
       return $connection->relevance_score[$node->ID];
    }
  }
]);
```

If that works is amazing.

When are you releasing a new version with the PR? Or better, how can I use it now?

**Comment by jasonbahl** [comment]
@a-barbieri I'll do a minor release right now 😄

**Comment by jasonbahl** [comment]
@a-barbieri sorry, got sidetracked last night. Will get it out today.

**Comment by jasonbahl** [comment]
@a-barbieri released:
https://github.com/wp-graphql/wp-graphql/releases/tag/v0.8.1

**Comment by jasonbahl** [comment]
@a-barbieri ya, I think you're on the right track. 

I'm not sure we have all the appropriate APIs in place to make this easy to register this edge to _all_ connections though. 

🤔 

I was thinking something like: 

```
$allowed_post_types = get_post_types( [ 'show_in_graphql' => true ], 'objects' );
if ( ! empty( $allowed_post_types ) ) {
  foreach ( $allowed_post_types as $allowed_post_type ) {
    register_graphql_field( 'RootQueryTo' . $allowed_post_type->graphql_single_name . 'ConnectionEdge', 'relevanceScore', $config );
  }
} 
```

But that only gets the `RootQuery->PostType` connections, and there might be many more connections than just root ones we'd want to apply the relevance score to. 

Ideally, you'd be able to maybe register fields like this to all connections that use the PostObjectLoader or something. That way anyone else registering new connections would get this added to that connection. 

🤔 

Need to think about this a bit. 

In the mean time, this should help you get further though.

```
// Get post types set to show in GraphQL
$allowed_post_types = get_post_types( [ 'show_in_graphql' => true ], 'objects' );

if ( ! empty( $allowed_post_types as $allowed_post_type ) ) {

  foreach ( $allowed_post_type as $allowed_post_type ) {
     register_graphql_field( 'RootQueryTo' .  $allowed_post_typ. 'ConnectionEdge', 'relevanceScore', $config );
  }

}
```

**Comment by a-barbieri** [comment]
> Ideally, you'd be able to maybe register fields like this to all connections that use the PostObjectLoader or something. That way anyone else registering new connections would get this added to that connection.

Yes, that would be ideal. Although I'm not sure it's needed. 

For example: when you search `blue` on `sofas`, `tables` and `lamps` you will might need to return a different set of fields for each, so you need three different queries. See below:

```
{
   sofas( where: { search: "blue" } ) {
      edges {
           relevanceScore
           node {
               textile
           }
       }
   }
   tables( where: { search: "blue" } ) {
      edges {
           relevanceScore
           node {
               material
           }
       }
   }
   lamps( where: { search: "blue" } ) {
      edges {
           relevanceScore
           node {
               dimmerable
           }
       }
   }    
}
```

You need relevance score because this way you can merge the 3 arrays and sort the results correctly.

```php
$allowed_post_types = get_post_types( [ 'show_in_graphql' => true ], 'objects' );
if ( ! empty( $allowed_post_types ) ) {
  foreach ( $allowed_post_types as $allowed_post_type ) {
    register_graphql_field( 'RootQueryTo' . $allowed_post_type->graphql_single_name . 'ConnectionEdge', 'relevanceScore', $config );
  }
} 
```

This would definitely help to release the plugin because you don't need to modify the custom post types manually. I'm not sure how to help here. What do you reckon?

**Comment by a-barbieri** [comment]
By the way the query currently requires `s` not `search` due to Relevanssi API. Maybe you want to update the posts above to avoid generating confusion if someone reads this thread in the future and falls into this little ambiguity.

**Comment by jasonbahl** [comment]
@a-barbieri we map `search` to `s` here: https://github.com/wp-graphql/wp-graphql/blob/develop/src/Data/Connection/PostObjectConnectionResolver.php#L390

This way WPGraphQL uses the more friendly `search` argument but still passes it to the underlying `WP_Query` as `s`

**Comment by a-barbieri** [comment]
Uh, that's great. Didn't know. It definitely reads better. I need to amend my code then.

**Comment by a-barbieri** [comment]
Ok, so I've updated the [gist](https://gist.github.com/a-barbieri/7802dee5938fee53e6f1a5cbd6f81cca) to use `search` instead of `s`. I've also updated the code to use WP Graphql `0.8.1`.

**Comment by a-barbieri** [comment]
Ok, so I tried dropping in this one using WP GraphQL `0.8.1`.

```php
add_action( 'graphql_register_types', 'register_relevance_score_field' );

function register_relevance_score_field() {
  // Prepare configuration to use later
  $config = [
              'type' => 'Float', // yes, it's a float
              'description' => __( 'Relevanssi score based on current search query', 'wp-graphql-relevanssi' ),
              'resolve' => function( $edge ) {
                $connection = isset( $edge['connection'] ) ? $edge['connection'] : null;
                $node = isset( $edge['node'] ) ? $edge['node'] : null;
                if ( is_null($connection) || is_null($node) ) {
                  return null;
                } else {
                   return $connection->relevance_score[$node->ID];
                }
              }
            ];

  // Append field to every connection edge
  $allowed_post_types = get_post_types( [ 'show_in_graphql' => true ], 'objects' );
  if ( ! empty( $allowed_post_types ) ) {
    foreach ( $allowed_post_types as $allowed_post_type ) {
      // I guess you what to make sure first letter is uppercase, correct?
      $graphql_single_name = ucfirst($allowed_post_type->graphql_single_name);
      // Register the `relevanceScore` field on the connection edge
      register_graphql_field( 'RootQueryTo' . $graphql_single_name . 'ConnectionEdge', 'relevanceScore', $config );
    }
  }
}
```

Unfortunately I don't see any `relevanceScore` on any edge the explorer. The function is executed: I tried inside a `add_action` callback as per [documentation](https://github.com/wp-graphql/wp-graphql/blob/f6d8c59fcf2bffa88fb30e222cbe1802764273a2/docs/source/extending/resolvers.mdx). 

I tried to debug by myself but there's no easy way to undestand what it does when we launch `register_graphql_field`. Do you have any suggestion?
